### PR TITLE
Create component standard working group

### PR DESCRIPTION
**Pull request by mtaufen**
As discussed in SIG-Architecture and KEP 0032:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

@luxas @sttts there are still a few TODO items, like setting our regular meeting time and agreeing on a name that isn't just the word "component." 

I'll hold off on running `make generate` and emailing stakeholder SIGS/steering committe with a link to this PR until we settle the name.

### CODE REVIEW

**Comment by mtaufen** [comment]
/hold
for stakeholders and steering approval

/sig architecture
/committee steering

File: wg-component-standard/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+* Regular WG Meeting: [Tuesdays at 08:30 PT (Pacific Time)](https://docs.google.com/document/d/18TsodX0fqQgViQ7HHUTAhiAwkf6bNhPXH4vNVTI7GwI) (weekly). [Convert to your timezone](http://www.thetimezoneconverter.com/?t=08:30&tz=PT%20%28Pacific%20Time%29).
+
+## Organizers
+
+* Lucas Käldström (**[@luxas](https://github.com/luxas)**), Luxas Labs (occasionally contracting for Weaveworks)
+* Dr. Stefan Schimanski (**[@sttts](https://github.com/sttts)**), Red Hat
+* Michael Taufen (**[@mtaufen](https://github.com/mtaufen)**), Google
+
+## Contact
+* [Slack](https://kubernetes.slack.com/messages/)
```
Comment (luxas) [inline wg-component-standard/README.md]: TODO for slack admins?

File: sigs.yaml [hunk 2]
```diff
@@ -2292,3 +2292,31 @@ workinggroups:
       url: https://docs.google.com/document/d/1RbC4SBZBlKth7IjYv_NaEpnmLGwMJ0ElpUOmsG-bdRA/edit
     contact:
       mailing_list: https://groups.google.com/forum/#!forum/kubernetes-wg-audit
+  - name: Component Standard
+    dir: wg-component-standard
+    mission_statement: >
+      Develop a standard foundation (philosophy and libraries) for core Kubernetes components to build on top of.
+      Areas to standardize include configuration (flags, ComponentConfig APIs, ...), status endpoints (healthz, configz, ...), integration points (delegated authn/z, ...), and logging.
+      Details are outlined in KEP 0032: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md.
+    charter_link:
```
Comment (luxas) [inline sigs.yaml]: Do we need this as a WG?

**Comment by pwittrock** [comment]
@luxas @mtaufen 

Couple questions:

- Working groups don't own code.  Instead code is owned by subprojects of SIGs.  Where (which SIg / subprojects) will the development be done out of?
- Working groups are designed to facilitate cross-SIG work with a clear beginning and end (after which the WG is disbanded).  Does this initiative have a clear end state?

Do we need this *and* a subproject for shared kubernetes infrastructure?  We also have a number of k8s libraries for things like sets and retries.  Do you envision these as being part of that initiative?

**Comment by mtaufen** [comment]
@DirectXMan12 @droot @deads2k @lavalamp

**Comment by mtaufen** [comment]
Offline chat with @pwittrock:
- Main concerns are ownership and clear end state.
- We should definitely find a SIG to own this repo and the building blocks we create. @luxas I saw you created the KEP under sig-cluster-lifecycle, but since this is about APIs, maybe sig-apimachinery is a better place since a lot of this relates to APIs?
- Regarding end state: 
  - The main reason to have a WG is that this discussion cuts across many sigs and we need a forum for that; sig-architecture is too high level to get into the nitty gritty details of how to implement these building blocks.
  - This should be a working-group at least until we can move the common code to a common place; at that point we can revisit the structure and decide how to collaborate going forward.

**Comment by mtaufen** [comment]
What is the difference between server-frameworks and server-sdk?

File: sigs.yaml [hunk 3]
```diff
@@ -2292,3 +2292,31 @@ workinggroups:
...
+        github: luxas
+        company: Luxas Labs (occasionally contracting for Weaveworks)
+      - name: Dr. Stefan Schimanski
+        github: sttts
+        company: Red Hat
+      - name: Michael Taufen
+        github: mtaufen
+        company: Google
+    meetings:
+    - description: Regular WG Meeting
```
Comment (luxas) [inline sigs.yaml]: We should add zoom link as soon as we create it as well

**Comment by luxas** [comment]
@pwittrock I'm happy to put this under `server-frameworks`. In large, this is a code split of the `k8s.io/apiserver` repo, along with newly-extracted code from `k8s.io/kubernetes`. This isn't an sdk, so it doesn't belong in `server-sdk`.

> The main reason to have a WG is that this discussion cuts across many sigs and we need a forum for that; sig-architecture is too high level to get into the nitty gritty details of how to implement these building blocks.
This should be a working-group at least until we can move the common code to a common place; at that point we can revisit the structure and decide how to collaborate going forward.

Agree on these. I'd also want to fill in that the end state is that our components behave the way we want them to (the WG will create a dedicated KEP for what standards we should have for our core components), and the code that made those standards happen are conveniently refactored into `k8s.io/component-base`

> @luxas I saw you created the KEP under sig-cluster-lifecycle, but since this is about APIs, maybe sig-apimachinery is a better place since a lot of this relates to APIs?

Yeah, just because I have authority to put things there and by the time I created the KEP initially I hadn't got SIG API Machinery/Arch approval (which we now have). Happy to move the KEP around, but think it could make sense to put this in a dedicated `wg-component-standard` folder as well...

For me, as said, I'm happy to put the code under `server-frameworks`, i.e. API Machinery.
That said, SIG Cluster Lifecycle will be part of this effort in various ways still.

> Do you envision these as being part of that initiative?

We envision this playing well together with all the other initiatives.

**Comment by mtaufen** [comment]
> I'd also want to fill in that the end state is that our components behave the way we want them to

I think @pwittrock's concern is that this is a moving target, potentially a perpetual motion machine :).
But for now a WG works for our purposes, and if it turns out this should be a longer lasting structure, we can revisit that later.

**Comment by pwittrock** [comment]
I don't intend to block this.  It is easy to unmake a WG, and I agree with the goals and execution.  I do want to probe a bit deeper - since the code will be done in a subproject of apimachinery anyway (proposed).  Why not just do the work out of the subproject.  It is fine for individuals to be working in multiple SIGs and for sub projects to have the collaboration tools of WG's.  The server-sdk project that owns kubebuilder is owned by @droot, @pwittrock (sig cli tl) and @DirectXMan12 (sig autoscaling).

Put another which SIGs **should** own the output of the WG, and **what value does the WG + subproject provide over just the subproject itself**?

**Comment by mtaufen** [comment]
@mattfarina 

> Who will own the repo k8s.io/component-base? A WG cannot own the code. If it is SIG API Machinery, shouldn't their ownership of the subproject be updated as well?

Yeah. It sounds like folks are leaning towards putting this in a SIG API Machinery sub-project. That could work, but I also want to make sure we have a public way of organizing this effort (regular open meetings, notes, etc.), because it broadly affects the project, and SIGs and WGs seem like the only two "official" ways to do this right now.

> A working group will help with cross SIG collaboration. What SIGs are planned on being apart of this? I ask to make sure they are looped in, announcements made at those SIGs, report out (using the new templates) a the community meeting for those SIGs should talk about it, etc.

I think this depends on each item. For example, ComponentConfig probably involves a mix of SIG Cluster Lifecycle and SIG API Machinery. Standardizing delegated auth implementation obviously involves SIG Auth. Standardizing endpoints like /configz, /healthz, etc. probably involves SIG Instrumentation. Our current plan is to write a KEP for each sub-area, and identify the relevant SIGs in those KEPs.

Thinking about this, maybe we want to separate "overall repo ownership" from ownership of the various pieces - which might align more directly with the involved SIGs for each sub-area.

> SIG Architecture has a new subproject for code organization. How does this relate to that?
Where will any conventions be documented for contributors and who will own that (A WG can't own these things)?

That's an interesting idea. Based on the description "Overall code organization, including github repositories and branching methodology, top-level and pkg OWNERS of kubernetes/kubernetes, vendoring," I'm not sure if it's a good fit, since our effort involves standardizing the implementation of various building-blocks, whereas the code organization subproject sounds higher-level (maybe it's not?).

> These are all logistical concerns for pulling this off effectively within the structure we have.

Yeah, totally agree these need to get figured out. Thanks for your clear list :)

@pwittrock

> I assume it is to get buy in for where the code is being moved from and that a WG facilitates this in some way?

Buy-in and transparent discussions, since it has a broad impact. Maybe KEPs are already good enough for that?

I'm definitely not against a sub-project, as there is less administrative work involved :P. But I want this work to be open and to our community's standard for collaboration.

**Comment by pwittrock** [comment]
Following up on a few points of clarification.

For collaboration:
In addition to a repo, you can get a Slack Channel, Email Discussion Group, Zoom Channel, Calendar Meeting, etc as a subproject.  Kubebuilder has a Slack channel that is relatively active and we have found to be a great way to collaborate.  It also has a email group that we used to setup a collocated BoF at Kubecon North America 2018.

A WG may be great if you have a large group of individuals that need to be informed, and a smaller group of individuals that are responsible.  The WG would give you a way to broadcast updates, but keep the subproject channels more focussed.  If the group of folks responsible is effectively the same as the group to be informed, the WG would probably just be the same folks within the subproject.

Regardless of if you have a WG or not, you will need a subproject to how the code and don't need to block on getting a WG to get a subproject.  You will need to get the support of the SIG TL's on the proposal KEP (e.g. @deads2k and @lavalamp for AM), and want to present it at a SIG meeting to be able to field questions.

**Comment by mtaufen** [comment]
@luxas @sttts @deads2k @lavalamp what do you guys think?
Shall we focus on making this a subproject of apimachinery for the time being?

**Comment by sttts** [comment]
> Shall we focus on making this a subproject of apimachinery for the time being?

If it helps to move forward, why not. 

I have my doubts though whether a subproject is the right tool, following @pwittrock characterization above. It's not about having a slack channels, email groups or zoom channels in the first place, but merely about refactoring existing code into a central place. If the efforts turn into a project of its own at some point, that's fine and we can create a formal subproject out of it. But for now in my view we have these goals:

1. move (existing) code to a central place (component-base repo) from places where they don't belong (mostly k8s.io/apiserver+apimachinery, some code in kubelet)
2. refactor that code to make it more reusable and refactor k/k components to use that shared code

Then, with a far far smaller priority:

3. make that code work for other projects in the ecosystem.

In other words, if Golang allowed to restrict the use of a repository to k/k components only, we would do that. Unfortunately it doesn't. Does this mean we should make this effort bigger than goal 1 and 2 for now, with design docs, a subproject, slack channel? IMO we shouldn't.

Let's get 1 and 2 done quickly (for 1.14+15), explicitly scope it down to those two, not the 3rd. And then later it should be feasible to turn the result into something bigger for wider adoption, possibly with further, larger refactorings, documentation, slack channel etc.

To avoid confusion and to avoid premature adoption, maybe we should reflect that in the repo or package names for now, or at the very least big and bold in the README.md. Until now it was enough to hide such code in k/k itself and claiming "we don't support it".

We need a sig as an owner for our new component-base repo (sig-arch, sig-api-machinery would both work IMO). But that's a different question. The work is cross-sig, through-out the whole code base. In that sense a working group probably is the right tool.

**Comment by sttts** [comment]
I have chatted with @luxas and @mtaufen about the "experimental" package idea I mentioned above to avoid the tension between the goals of
1. building something for the wider ecosystem to consume
2. versus moving fast with refactoring the vast amount of code we have out into component-base and re-using it in our in-tree k/k component.

The former will certainly require follow-up KEPs for some of the work (further generalization of componentconfigs comes to mind). The later should improve code quality in 1.14 and 1.15 and would make componentconfig implementation inside k/k feasible in the first place. The later is also about learning what we need in order to reach the first goal. 

In order to allow rapid refactoring for the second goal, we propose to

1. create the component-base repository as soon as possible
2. but move unfinished, factored out code into `k8s.io/component-base/experimental/<sub-package>` first.

We will then promote packages to the top-level k8s.io/component-base/<sub-package> when we think they are ready for public consumption.

We will try to minimize the use of those experimental packages in k8s.io/apiserver (which is already used by 3rdparties) or at best avoid it completely. On the other hand, kube-controller-manager, kube-scheduler, kubelet, kube-apiserver are all in-tree completely and can serve as test beds while the new code becomes mature and ready for promotion.

That having said, the working group should serve goal 2 and prepare for goal 1. @pwittrock does that match your conception of WG vs. sub-project?

**Comment by luxas** [comment]
Hi all, are there still outstanding issues/comments/concerns or can we proceed with the WG creation?

Simply put, we want WG + API Machinery subproject because:
a) The WG will work on cross-cutting issues wrt Component Standards (exactly as the name suggests), and will focus on the "docs-first" approach and develop KEPs wrt how a k8s component should work, as well as how to graduate existing (or non-existent to date) ComponentConfig schemas to beta and higher, in collaboration with SIG Architecture and others.
b) The `k8s.io/component-base` subproject will focus on moving code out from code in the proposed and approved fashion ("code-first", if you like). The WG will help keep the subproject on the right track, and involve the right stakeholders when more complex changes are made (essentially help with communication to the rest of the community)

If this makes sense, can we please proceed with this? We intend all the best for the project and if it later turns out this way of working wasn't the ideal one, it is easy to unmake the WG later, as you said.

@pwittrock @mattfarina @thockin does this make sense?

**Comment by luxas** [comment]
@mtaufen when updating this with the final feedback, please include `k8s.io/component-base` as a AM subproject as per https://github.com/kubernetes/org/issues/330#issuecomment-452068510

**Comment by mtaufen** [comment]
@luxas @sttts I updated this to add the component-base subproject to sig-api-machinery.

Are there any additional updates folks would like to see before we merge this?
Are we set on Tuesday at 8:30AM PST as the meeting time?

File: sigs.yaml [hunk 4]
```diff
@@ -2340,3 +2344,31 @@ workinggroups:
...
+      day: Tuesday
+      time: "08:30"
+      tz: "PT (Pacific Time)"
+      frequency: weekly
+      url: https://docs.google.com/document/d/18TsodX0fqQgViQ7HHUTAhiAwkf6bNhPXH4vNVTI7GwI
+    contact:
+      mailing_list: https://groups.google.com/forum/#!forum/kubernetes-wg-component-standard
```
Comment (luxas) [inline sigs.yaml]: Follow-ups: add `slack` (#wg-component-standard, once created), and `teams` (@kubernetes/wg-component-standard, once created)

File: sigs.yaml (hunk 2 above)
Comment (luxas) [reply sigs.yaml]: you could also omit this from the YAML to not confuse anybody

File: sigs.yaml [hunk 5]
```diff
@@ -2340,3 +2344,31 @@ workinggroups:
     contact:
       slack: wg-security-audit
       mailing_list: https://groups.google.com/forum/#!forum/kubernetes-wg-audit
+  - name: Component Standard
+    dir: wg-component-standard
```
Comment (luxas) [inline sigs.yaml]: consider also adding `label: wg/component-standard` here. can be followed up

**Comment by luxas** [comment]
> Are we set on Tuesday at 8:30AM PST as the meeting time?

Yep, I think so

> Are there any additional updates folks would like to see before we merge this?

I left comments but please consider them non-blocking. I think we can merge now.

/hold cancel
/lgtm

Please hit the `/approve` button @thockin @bgrant0607 or @jdumars :). Thanks!

**Comment by dims** [comment]
LGTM 👍

**Pull request merged**